### PR TITLE
IAction.Render() & IAction.Unrender()

### PR DESCRIPTION
**Pull request by dahlia**
This patch implements a minimum viable subset of the idea I had suggested in https://github.com/planetarium/libplanet/issues/31#issuecomment-481699147.

There are missing behaviors yet; `Render()` does not guarantee it's called after the action is confirmed in the network (we need to define what “confirmation” means here), etc, for instance.  The currently guaranteed attributes are:

 -  `IAction.Render()` and `IAction.Unrender()` take an equivalent `IActionContext` object with a clean state (which means its `Random` state is never consumed) to `IAction.Execute()` took.
 -  If some blocks become “stale,” `Unrender()` methods of actions in these blocks are invoked in reverse order.
 -  For each action, `Render()` is only once called unless it's stale so that its `Unrender()` is called.
 -  For each action, the number of `Unrender()` calls is the same to or one less than the number of `Render()` calls.

### CODE REVIEW

File: Libplanet/Blocks/Block.cs [hunk 1]
```diff
@@ -150,6 +150,52 @@ public byte[] ToBencodex(bool hash, bool transactionData)
...
+        /// the <see cref="Transactions"/> and their
+        /// <see cref="Transaction{T}.Actions"/> (e.g., tx&#xb9;-act&#xb9;,
+        /// tx&#xb9;-act&#xb2;, tx&#xb2;-act&#xb9;, tx&#xb2;-act&#xb2;,
+        /// &#x2026;).
+        /// Note that each <see cref="IActionContext.Random"/> object has
+        /// a unconsumed state.
+        /// </returns>
+        [Pure]
+        public
+        IEnumerable<(Transaction<T>, T, IActionContext, IAccountStateDelta)>
```
Comment (longfin) [inline Libplanet/Blocks/Block.cs]: Quadruples seems to be a bit complicated as a return type for public methods. How about creating a type?

File: Libplanet/Tx/Transaction.cs [hunk 2]
```diff
@@ -460,31 +466,40 @@ public byte[] ToBencodex(bool sign)
         /// <c>false</c>.
         /// </exception>
         [Pure]
-        public IAccountStateDelta EvaluateActions(
+        public IEnumerable<(T, IActionContext, IAccountStateDelta)>
```
Comment (longfin) [inline Libplanet/Tx/Transaction.cs]: This triples also seems to be a bit complicated.

File: Libplanet/Tx/Transaction.cs [hunk 3]
```diff
@@ -460,31 +466,40 @@ public byte[] ToBencodex(bool sign)
...
+            ActionContext CreateActionContext(
+                IAccountStateDelta prevStates,
+                int randomSeed
+            ) =>
+                new ActionContext(
+                    signer: Signer,
+                    miner: minerAddress,
+                    blockIndex: blockIndex,
+                    previousStates: prevStates,
+                    randomSeed: randomSeed,
```
Comment (longfin) [inline Libplanet/Tx/Transaction.cs]: Is it okay not to have `unchecked`?

File: Libplanet/Blockchain/BlockChain.cs [hunk 4]
```diff
@@ -468,6 +511,30 @@ internal BlockLocator GetBlockLocator(int threshold = 10)
         // we need to add a synchronization mechanism to handle this correctly.
         internal void Swap(BlockChain<T> other)
         {
+            // Finds the branch point.
+            Block<T> topmostCommon = null;
+            for (long i = Math.Min(other.Tip.Index, Tip.Index); i >= 0; i--)
```
Comment (longfin) [inline Libplanet/Blockchain/BlockChain.cs]: We can use `Block<T>.PreviousHash` to avoid index lookup.

File: Libplanet.Tests/Blocks/BlockTest.cs [hunk 5]
```diff
@@ -208,35 +246,70 @@ public void EvaluateActions()
...
+                .ToImmutableArray();
+            i = 0;
+            expectations = new[]
+            {
+                (0, 0, new[] { "A,D", "B", "C", null, null }),
+                (1, 0, new[] { "A,D", "B", "C", "E", null }),
+                (2, 0, new[] { "A,D", "B", "C", "E", "RecordRehearsal:False" }),
+            };
+            Assert.Equal(expectations.Length, tuples.Length);
+            foreach (var tuple in tuples)
```
Comment (earlbread) [inline Libplanet.Tests/Blocks/BlockTest.cs]: How about using `for` loop if you use `i`?

File: Libplanet.Tests/Blocks/BlockTest.cs [hunk 6]
```diff
@@ -208,35 +246,70 @@ public void EvaluateActions()
...
+            {
+                var expect = expectations[i++];
```
Comment (earlbread) [inline Libplanet.Tests/Blocks/BlockTest.cs]: Rider says we can use deconstruction like:
```csharp
var (transaction, dumbAction, actionContext, accountStateDelta) = tuples[i];
var (item1, item2, strings) = expectations[i];
```
How about using deconstruction?
It looks more readable.

File: Libplanet.Tests/Common/Action/DumbAction.cs [hunk 7]
```diff
@@ -62,16 +87,113 @@ public IAccountStateDelta Execute(IActionContext context)
...
             RecordRehearsal = (bool)plainValue["record_rehearsal"];
+            RecordRandom =
+                plainValue.ContainsKey("record_random") &&
+                plainValue["record_random"] is bool r &&
+                r;
+        }
+
+        public bool Equals(DumbAction other)
+        {
+            return !ReferenceEquals(null, other) && (
```
Comment (earlbread) [inline Libplanet.Tests/Common/Action/DumbAction.cs]: Is this different from `!(other is null)`?

Comment (dahlia) [reply Libplanet.Tests/Common/Action/DumbAction.cs]: No, it means the same.  It's generated by Rider.  I will changed it to use pattern matching syntax.

File: Libplanet/Blocks/Block.cs (hunk 1 above)
Comment (dahlia) [reply Libplanet/Blocks/Block.cs]: Fixed: 19089e75db6287a8e9de6eb47b92f5d72eef0d1d.

File: Libplanet/Tx/Transaction.cs (hunk 2 above)
Comment (dahlia) [reply Libplanet/Tx/Transaction.cs]: Fixed: 19089e75db6287a8e9de6eb47b92f5d72eef0d1d.

File: Libplanet.Tests/Blocks/BlockTest.cs (hunk 6 above)
Comment (dahlia) [reply Libplanet.Tests/Blocks/BlockTest.cs]: Fixed by adding a distinct type: 19089e75db6287a8e9de6eb47b92f5d72eef0d1d.

File: Libplanet.Tests/Common/Action/DumbAction.cs (hunk 7 above)
Comment (dahlia) [reply Libplanet.Tests/Common/Action/DumbAction.cs]: Fixed: 10f136b87996339da9678896db76bdfa6ead6807.

File: Libplanet.Tests/Blocks/BlockTest.cs (hunk 5 above)
Comment (dahlia) [reply Libplanet.Tests/Blocks/BlockTest.cs]: Fixed by using `Enumerable.Zip()`: 78faaa43e34ec37cebf0793e13e7d30db92ca11d.

File: Libplanet/Blockchain/BlockChain.cs (hunk 4 above)
Comment (dahlia) [reply Libplanet/Blockchain/BlockChain.cs]: Fixed: 9db8f48c4dd539b17559cf3723a9703f28cbeedd.

File: Libplanet/Tx/Transaction.cs (hunk 3 above)
Comment (dahlia) [reply Libplanet/Tx/Transaction.cs]: Fixed: 604f9b57060498eecd3c03783be742c39f0260af.

**Comment by dahlia** [comment]
@earlbread @longfin I rebased this on the current master and addressed review feedback.  Could you review this again?

File: CHANGES.md [hunk 8]
```diff
@@ -70,18 +71,30 @@ To be released.
...
 [#205]: https://github.com/planetarium/libplanet/pull/205
 [#206]: https://github.com/planetarium/libplanet/pull/206
 [#210]: https://github.com/planetarium/libplanet/pull/210
 [#213]: https://github.com/planetarium/libplanet/pull/213
 [#215]: https://github.com/planetarium/libplanet/pull/215
+<<<<<<< HEAD
 [Ethereum Homestead algorithm]: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md
+=======
+[#216]: https://github.com/planetarium/libplanet/pull/216
+>>>>>>> Format/writing adjustments on CHANGES.md
```
[the remark this answers is not in the document]
Comment (dahlia) [reply CHANGES.md]: Fixed!

**Comment by dahlia** [comment]
@longfin Please review this again.

**Pull request merged**